### PR TITLE
Align MAE pretrain

### PR DESCRIPTION
**Pull request by leaves-zwx**

### CODE REVIEW

File: libai/models/utils/graph_base.py [hunk 1]
```diff
@@ -48,7 +48,7 @@ def __init__(
             if fp16:
                 self.config.enable_amp(True)
                 grad_scaler = flow.amp.GradScaler(
-                    init_scale=65536.0,
+                    init_scale=65536.0 * dist.get_data_parallel_size(),
```
Comment (leaves-zwx) [inline libai/models/utils/graph_base.py:51]: 虽然与 pytorch 对齐的 init scale 默认值是 65536，但在 global 下相对于 pytorch 的 local，loss.mean() 会多除1个 split num（纯数据并行下就是 device num），导致反向开始时，init grad 也会多除1个 split num，整个反向计算过程中 grad 都会小 split num 倍这个量级。

虽然 amp.GradScaler 会动态调整 scale 来使 grad 保持在合理的精度范围内（一般就是 float16 的表示范围），但这里乘上 data parallel size 可以使在第一轮迭代时，grad 补偿性的放大，可以让其在一开始就与 pytorch 的计算方式保持一致。

这个特性应该不会影响模型训练最终收敛，但在调试模型的收敛，算法的一致性时应该能起到一定的帮助。

File: projects/MAE/configs/mae_finetune.py [hunk 2]
```diff
@@ -14,11 +14,11 @@
 # limitations under the License.
 
 from omegaconf import OmegaConf
-
 from flowvision.data import Mixup
-from flowvision.loss.cross_entropy import SoftTargetCrossEntropy
 
+# from flowvision.loss.cross_entropy import SoftTargetCrossEntropy
 from libai.config import LazyCall, get_config
+from modeling.cross_entropy import SoftTargetCrossEntropy
```
Comment (leaves-zwx) [inline projects/MAE/configs/mae_finetune.py:21]: 用 MAE 自定义的 SoftTargetCrossEntropy，因为需要对其进行 hack，来使 mean 和 sum 保持在 float32 下计算。

File: projects/MAE/configs/models/mae_vit_base_patch16.py [hunk 3]
```diff
@@ -18,6 +18,6 @@
     decoder_num_heads=16,
     mlp_ratio=4,
     norm_layer=partial(LayerNorm, eps=1e-6),
-    norm_pix_loss=False,
+    norm_pix_loss=True,
```
Comment (leaves-zwx) [inline projects/MAE/configs/models/mae_vit_base_patch16.py:21]: 参考 https://github.com/facebookresearch/mae/blob/efb2a8062c206524e35e47d04501ed4f544c0ae8/PRETRAIN.md?plain=1#L11

File: projects/MAE/modeling/cross_entropy.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+        super(SoftTargetCrossEntropy, self).__init__()
+
+    def forward(self, x: flow.Tensor, target: flow.Tensor) -> flow.Tensor:
+        pred = flow.log_softmax(x, dim=-1)
+        loss = -target * pred
+        # sum and mean should be calculated with float32
+        # amp_white_identity ensure -target * pred using float16
+        # amp_black_identity ensure sum and mean using float32
+        loss = flow._C.amp_white_identity(loss)
+        loss = flow._C.amp_black_identity(loss)
```
Comment (leaves-zwx) [inline projects/MAE/modeling/cross_entropy.py:31]: 为了在这里控制后面 sum 和 mean 都用 float32 计算

File: projects/MAE/utils/weight_convert.py [hunk 5]
```diff
@@ -63,7 +63,9 @@ def filter_keys(key, value, cfg):
     """
     Filtering the state_dict keys and values to match LiBai's MAE model
     """
-    if "norm1" in key:
+    if key.startswith("decoder_"):
```
Comment (leaves-zwx) [inline projects/MAE/utils/weight_convert.py:66]: finetune 加载 pretrain 模型时，忽略其中的 decoder 部分，只训 encoder。

File: libai/models/utils/graph_base.py (hunk 1 above)
Comment (Ldpe2G) [reply libai/models/utils/graph_base.py:51]: global 的 数据并行相比 eager ddp， loss.mean 会多除一个 world_size，这里对Loss做scale，乘以了一个 world_size，反向，就和 eager ddp 一致了。
但是 eager ddp 在 grad allreduce 之后会除以一个  world_size

感觉global 模式下要对齐 eager ddp, 是不是 loss 要乘以 world_size，然后更新参数前 grad 手除以 world_size？

File: projects/MAE/modeling/cross_entropy.py (hunk 4 above)
Comment (Ldpe2G) [reply projects/MAE/modeling/cross_entropy.py:31]: 这里这个 amp__xx__identity 接口的设计，感觉有点让人迷惑，两个连着调用
感觉也是得知道实现细节的才知道怎么用，这个是不打算给用户用的是吧？

Comment (strint) [reply projects/MAE/modeling/cross_entropy.py:31]: 这个一直没改，可以现在把这个接口改了，改成和算法无关的

File: libai/models/utils/graph_base.py (hunk 1 above)
Comment (leaves-zwx) [reply libai/models/utils/graph_base.py:51]: 这个 feature 跟 graph 没有绑定关系，eager 下也应该要实现的。

https://pytorch.org/docs/stable/amp.html#gradient-scaling 

GradScaler 的作用方式，在 loss.backward() 前对其乘1个动态 scale 系数，反向开始时，初始梯度也会乘这个 scale 系数（反向传播原理），这样的目的在于控制反向传播的过程中，梯度能维持在 float16 所能表示的精度范围内。在反向传播到参数更新前一步的时候，需要把梯度从 float16 转换到 float32，这时再把梯度除这个 scale 系数，其实在数学计算上等价，而这时精度已经恢复到 float32，不用担心溢出。

而在 global 中，我们做 loss.mean().backward() 时，可以认为 mean 比 pytorch 多除了 world size 倍（数据并行），但这里与上面的系数是两个独立的概念，只是因为乘除法的结合律可以合到一起乘/除。

ddp 下 loss.mean().backward() 因为单卡视角，不会多除 world size，所以没这个问题。但最后从计算原理来讲，梯度 allreduce 后需要对梯度再除 world size 才是正确的计算方式。可以认为 global 的单机视角才是正确的梯度计算方式，ddp 才是一种 trick 的实现，它在梯度开始传播时少了 world size 倍，在最后更新前才补偿回来。

Comment (leaves-zwx) [reply libai/models/utils/graph_base.py:51]: ddp 下 amp 也需要  GradScaler，但 init scale 就不用乘那个 world size 了。

File: projects/MAE/modeling/cross_entropy.py (hunk 4 above)
Comment (hjchen2) [reply projects/MAE/modeling/cross_entropy.py:31]: 要不支持一下autocast的用法，只是graph和eager下autocast的行为不一样，lazy下autocast只是标记某个op是否可以自动cast

Comment (leaves-zwx) [reply projects/MAE/modeling/cross_entropy.py:31]: 没太理解houjiang说的：

- graph和eager下autocast的行为不一样：这里有什么不一样，以及为什么要实现成不一样的呢？
- lazy下autocast只是标记某个op是否可以自动cast：这里能解决这里 trick 的问题吗？

我理解这里 trick 的本质是：lazy amp 没有推导出我们理想的精度设置，所以需要人工干预，而人工干预又要顾虑 lazy amp 的特性：它会往前搜，它不会尊重 flow.cast 等用户意志。

lazy amp 为什么没有推导出我们理想的精度设置呢？因为其只考虑前驱后继的影响，但在深度学习实际业务中，后向也会受前向的精度影响。算法设计上没把这个业务知识考虑进去。比如前向中的 reduce sum 我们一般认为其是 black，但 broadcast add 展开的后向 reduce sum 一般认为其是 white。不能光考虑黑白名单。

Comment (leaves-zwx) [reply projects/MAE/modeling/cross_entropy.py:31]: lazy amp 算法拿易用性换一点点潜在的自动优化可能，我个人觉得是不划算的。如果真有潜在优化的可能，不依靠自动，靠手动达到也行，这里的自动意义有多大，我是比较怀疑的。

**Review by Ldpe2G — APPROVED** [review]

File: projects/MAE/configs/mae_pretraining.py [hunk 6]
```diff
@@ -42,33 +46,60 @@
...
+
+# enable activation checkpointing
+# train.activation_checkpoint.enabled = True
+
+# set rdma enabled when num nodes > 1
+# train.rdma_enabled = False
+
+effective_batch_size = train.train_micro_batch_size * train.num_accumulation_steps * n_gpus
+epoch_iter = dataset_train_length // effective_batch_size
+train.checkpointer.period = epoch_iter * 20
```
Comment (CPFLAME) [inline projects/MAE/configs/mae_pretraining.py:67]: 我看上面定义了一个`dataset_train_length`, 看上去是人眼计算出来的数据集长度, 如果换一个数据集可能又需要重新计算, 
所以我觉得这里是可以直接在`projects/MAE/train_net.py`下面直接重载一下这个函数 https://github.com/Oneflow-Inc/libai/blob/a3b38c1a3fdac67016c23876172b0aadcca092b2/libai/engine/default.py#L678-L731

可以直接在`config.py`里面加一个类似`train.save_model_after_n_epoch=20`之类的参数, 然后在`train_net.py`下面重载一下, 这样就不用每次都去数数据集的长度, 然后定义在config下面了.

```python
    @classmethod
    def auto_scale_hyperparams(cls, cfg, data_loader):
        super().auto_scale_hyperparams(cfg, data_loader)
        iter_per_epoch = len(data_loader.dataset) // cfg.train.global_batch_size
        cfg.train.checkpointer.period = iter_per_epoch * cfg.train.save_model_after_n_epoch
```

**Review by CPFLAME — APPROVED** [review]

File: projects/MAE/modeling/cross_entropy.py (hunk 4 above)
Comment (chengtbf) [reply projects/MAE/modeling/cross_entropy.py:31]: 那更加合适的解决方案是什么呢？ @leaves-zwx  😂 

#### 原版：

```python
def forward(self, x: flow.Tensor, target: flow.Tensor) -> flow.Tensor:
        pred = flow.log_softmax(x, dim=-1)
        loss = -target * pred
        # sum and mean should be calculated with float32
        # amp_white_identity ensure -target * pred using float16
        # amp_black_identity ensure sum and mean using float32
        loss = flow._C.amp_white_identity(loss)  # 这两行出现在代码里，用户是看不懂的。至少要改一个接口
        loss = flow._C.amp_black_identity(loss)
        loss = flow.sum(loss, dim=-1)
        return loss.mean()
```


#### 改进版讨论

> 要不支持一下autocast的用法，只是graph和eager下autocast的行为不一样，lazy下autocast只是标记某个op是否可以自动cast

后江说的这个方案，是不是用这种方式来代替：

```python
def forward(self, x: flow.Tensor, target: flow.Tensor) -> flow.Tensor:
        pred = flow.log_softmax(x, dim=-1)
        loss = -target * pred
        with autocast(enabled=False): # 语义上是关闭 amp
            loss = flow.sum(loss.float(), dim=-1) # 手动让 sum 输入是 fp32
            loss = loss.mean()
        return loss
```

参考 pytorch 链接： https://pytorch.org/docs/stable/amp.html#torch.autocast

以及 pytorch 官网示例代码：

```python
# Creates some tensors in default dtype (here assumed to be float32)
a_float32 = torch.rand((8, 8), device="cuda")
b_float32 = torch.rand((8, 8), device="cuda")
c_float32 = torch.rand((8, 8), device="cuda")
d_float32 = torch.rand((8, 8), device="cuda")

with autocast():
    e_float16 = torch.mm(a_float32, b_float32)
    with autocast(enabled=False):
        # Calls e_float16.float() to ensure float32 execution
        # (necessary because e_float16 was created in an autocasted region)
        f_float32 = torch.mm(c_float32, e_float16.float())

    # No manual casts are required when re-entering the autocast-enabled region.
    # torch.mm again runs in float16 and produces float16 output, regardless of input types.
    g_float16 = torch.mm(d_float32, f_float32)
```
